### PR TITLE
Revise TCK Process 1.0 to include patch updates to TCKs

### PR DESCRIPTION
**Pull request by scottmarlow**
Signed-off-by: Scott Marlow <smarlow@redhat.com>
https://github.com/jakartaee/specification-committee/issues/45

### CODE REVIEW

**Comment by scottmarlow** [comment]
Need a review by those on jakarta.ee-spec@eclipse.org, sending email now asking for a review...

**Comment by scottmarlow** [comment]
[Review link of tckprocess document.](https://deploy-preview-1018--jakartaee.netlify.app/committees/specification/tckprocess/)

**Comment by scottmarlow** [comment]
Question for reviewers of this change, do we need to increment the `1.0` version with this change?

**Comment by chrisguindon** [comment]
@ivargrimstad who should we add to this issue to help answer questions from @scottmarlow

**Comment by ivargrimstad** [comment]
@chrisguindon These process documents are owned by the specification committee. I suggest we get it on the agenda for the next spec committee call. I'll take the action to do so.

**Comment by aeiras** [comment]
Hola @chrisguindon, 

Please submit your questions, the issue or PR directly to the TCK public forum by using  jakartaee-tck-dev@eclipse.org. 

The TCK mailing list is active and quite eager to get stuff moving, with 96 subscribers you can trust me when I tell you that we will tackle anything that arrives to that forum.

**Comment by scottmarlow** [comment]
I just sent mail to jakartaee-tck-dev@eclipse.org asking for feedback/review on this pr.  Thanks for the suggestion @aeiras!

File: content/committees/specification/tckprocess/index.md
Comment (scottmarlow) [inline content/committees/specification/tckprocess/index.md:428]: We are releasing:
- Jakarta EE Platform TCK 9.0.1
- Jakarta Servlet TCK 5.0.1

I will follow the details mentioned in this pull request in case it gets merged in the future.

**Comment by scottmarlow** [comment]
@kwsutter please review, I don't think I can add you as a reviewer.

Note, one more change is now expected for TCK challenges.

**Comment by scottmarlow** [comment]
I'm going to push a new change...

**Comment by kwsutter** [comment]
> Question for reviewers of this change, do we need to increment the `1.0` version with this change?

Yes.  Based on the extent of the changes, I would suggest a 1.1 version.

File: content/committees/specification/tckprocess/index.md
Comment (kwsutter) [inline content/committees/specification/tckprocess/index.md:28]: Are we going to differentiate between a minor point release (x.y) versus a service release (x.y.z)?  Or, will the process be the same?

File: content/committees/specification/tckprocess/index.md
Comment (kwsutter) [inline content/committees/specification/tckprocess/index.md:223]: This sounds a little backwards.  The TCK challenge can suggest anything they want.  But, the real driver here is the owner of the TCK.  So, if the TCK owner determines that the challenge is valid and they document a workaround of using a specific Jakarta deployment descriptor, then those tests would not have to be excluded.  We need the TCK owner to determine what are valid workarounds.

File: content/committees/specification/tckprocess/index.md
Comment (kwsutter) [inline content/committees/specification/tckprocess/index.md:253]: Define "major".  Wouldn't we also expect that the exclude lists get emptied on a minor (x.y) release?  We would not expect this for a service (x.y.z) release, but I think we should expect it for both major and minor releases.

File: content/committees/specification/tckprocess/index.md
Comment (kwsutter) [inline content/committees/specification/tckprocess/index.md:432]: Let's pretend that this section also applies to service releases (x.y.z).  When is it required for an implementation to move up to the latest x.y.z version of the TCK?  Maybe that's spelled out elsewhere in the TCK process.  I'm only reviewing the updates right now.  But, I thought we had some grace period defined to allow an organization to continue using the previous versions of the TCK up to a certain point in time.  This allows an existing test effort to continue to completion without having to start over with a new TCK revision.

**Review by kwsutter — COMMENTED** [review]
> @kwsutter please review, I don't think I can add you as a reviewer.

Yeah, one of the shortcomings with storing these documents on jakarta.ee.  Only certain people have the permissions to be reviewers, approvers, and mergers.  Calling me out via a comment has worked on other similar issues and PRs.

> 
> One question is whether we should also update the TCK process flowchart as well.

Might have to.  We're adding clarifications to the TCK release process.  If updating the flowchart is a major pain, then you'll have to evaluate whether the general flow is still covered or not.  Or, whether we will have to eat the cost and figure out how to update the flowchart.

File: content/committees/specification/tckprocess/index.md
Comment (scottmarlow) [reply content/committees/specification/tckprocess/index.md]: From a TCK versioning perspective, we always use x.y.z (e.g. 5.0.1 for Servlet TCK or 9.0.1 for Platform TCK).  We kind of know that and aren't questioning that.  Just mentioning for those following this discussion and wondering which version the x.y vs x.y.z refers to.

My intention with using the words that I did in `Process for releasing a point revision of a TCK` was to cover what we do when addressing TCK challenges with a TCK x.y.z release with the challenged tests made to pass via test exclusion (or other workaround to be mentioned).

Perhaps the intention would be clearer with different wording `Process for (TCK) service release (x.y.z) to resolve a TCK challenge`.

> Are we going to differentiate between a minor point release (x.y) versus a service release (x.y.z)? Or, will the process be the same?

I didn't intend this new section to cover what happens for a minor or major EE point release but I'm not 100% sure if you are referring to the TCK or SPEC or EE versioning?

File: content/committees/specification/tckprocess/index.md
Comment (scottmarlow) [reply content/committees/specification/tckprocess/index.md]: Okay, if we loosen up what are valid workarounds that is more open ended but works equally as long as the workaround is not a `test code` change.

Comment (scottmarlow) [reply content/committees/specification/tckprocess/index.md]: We could rephrase as:
`The specification project may approve workarounds for an accepted TCK challenge so that the tests do not need to be excluded.`  

This would also avoid the need for a new TCK release.

File: content/committees/specification/tckprocess/index.md
Comment (scottmarlow) [reply content/committees/specification/tckprocess/index.md:252]: Some possible options:
1. Only clear the excluded test list for specifications that are being updated/developed for a major/minor Jakarta EE release.  The burden is on the EE implementation teams to challenge any of those tests that aren't fixed (which slows adoption of that new EE version.)
2. Always clear the excluded test list for specifications that are included in any Jakarta EE minor/major release but not service (x.y.z) releases.

Would it make sense to remove all tests that are repeatedly challenged but not fixed for two consecutive EE major/minor releases?

File: content/committees/specification/tckprocess/index.md
Comment (scottmarlow) [reply content/committees/specification/tckprocess/index.md:435]: > Let's pretend that this section also applies to service releases (x.y.z). When is it required for an implementation to move up to the latest x.y.z version of the TCK? 

I have been thinking about how this is undefined for Jakarta EE and we should answer this question.  The cost of continually being able to process TCK challenges for every released EE version forever, is too great.  

> Maybe that's spelled out elsewhere in the TCK process. 

It is not spelled out anywhere that I am aware of.

> I'm only reviewing the updates right now. But, I thought we had some grace period defined to allow an organization to continue using the previous versions of the TCK up to a certain point in time. 

I haven't heard about that for Jakarta EE.  

> This allows an existing test effort to continue to completion without having to start over with a new TCK revision.

Are you saying that they could continue to use the Jakarta EE 9.0.0 Platform TCK forever without having to start over with the latest 9.0.x Platform TCK version?

**Comment by scottmarlow** [comment]
> > Question for reviewers of this change, do we need to increment the `1.0` version with this change?
> 
> Yes. Based on the extent of the changes, I would suggest a 1.1 version.

Thanks @kwsutter, will do that!

**Comment by scottmarlow** [comment]
> > One question is whether we should also update the TCK process flowchart as well.
> 
> Might have to. We're adding clarifications to the TCK release process. If updating the flowchart is a major pain, then you'll have to evaluate whether the general flow is still covered or not. Or, whether we will have to eat the cost and figure out how to update the flowchart.

I installed https://wiki.gnome.org/Apps/Dia/Download which does let me make changes to the jakarta.ee/content/committees/specification/tckprocess/TCKProcess.png image file but I wonder if the png file was generated from a different program for which we have some other input file for elsewhere.

@chrisguindon do you know if https://github.com/jakartaee/jakarta.ee/commits/src/content/committees/specification/tckprocess/TCKProcess.png is the right file for us to make changes to?  Or is there another file elsewhere that we should edit?

File: content/committees/specification/tckprocess/index.md
Comment (scottmarlow) [reply content/committees/specification/tckprocess/index.md:435]: We should decide soon how far back we support TCK challenges for, so that it is clear when we require moving up to the latest TCK revision.

File: content/committees/specification/tckprocess/index.md
Comment (scottmarlow) [reply content/committees/specification/tckprocess/index.md]: @kwsutter I just pushed an update to address this by referring to service releases.

File: content/committees/specification/tckprocess/index.md
Comment (scottmarlow) [reply content/committees/specification/tckprocess/index.md]: @kwsutter I also updated the wording for this.

File: content/committees/specification/tckprocess/index.md
Comment (kwsutter) [reply content/committees/specification/tckprocess/index.md:252]: I actually like your last idea.  If component specs can't step up to fixing some challenged (and accepted) tests in a timely manner, then just remove them.  This would help resolve the issue we've had with the JSF signature tests for several releases...

But, we would need a strong requirement statement on fixing them as first priority.  Thus, I like your option #2 as the path for component TCKs to follow.  And, then if these are not resolved in 2 or 3 releases, then the tests get removed.

File: content/committees/specification/tckprocess/index.md
Comment (kwsutter) [reply content/committees/specification/tckprocess/index.md:435]: I can no longer find that defined "grace period" for running a given TCK.  But, I did notice that this TCK Process overlaps with the TCK User's Guide in a few areas, especially with the challenge process.  How do we keep these documents in sync?  Or, do we just declare this document as "gospel" and then point to this from the TCK User's Guide?

Getting back to the "grace period"...  We state that the TCK must match the major.minor version of the Spec.  Maybe that's sufficient.  Test efforts can use whatever x.y.z version of the TCK (ie. any service release) to test a compatible implementation.

**Review by kwsutter — APPROVED** [review]
Thanks, @scottmarlow, for the updates!  Looks good from my viewpoint.

**Review by kwsutter — APPROVED** [review]
This looks good.  I see that the version of this TCK Process document is already at 1.1?  How is this being managed?  Are we collecting updates to the process and then doing a minor version update at some point in time?  Or, is every update to the process causing an update to the minor version (or maybe a service release to the TCK process)?  I'll let you figure that part out.  But, the specific change for the Signature Tests looks good.  Thanks!

**Comment by scottmarlow** [comment]
> This looks good. I see that the version of this TCK Process document is already at 1.1? 

This pull request is incrementing the TCK Process document from `1.0` to `1.1` but we haven't really defined the requirements for updating the versioning of the `TCK Process`.  This seems like a good topic for the Spec mailing list.

> How is this being managed? Are we collecting updates to the process and then doing a minor version update at some point in time? Or, is every update to the process causing an update to the minor version (or maybe a service release to the TCK process)? 

These are also good questions as we didn't discuss any of this as well.

> I'll let you figure that part out. But, the specific change for the Signature Tests looks good. Thanks!

:+1: 

@kwsutter I made one more change to cover the possible TCK changes that we might later make for the removal of the Java Security Manager via https://github.com/jakartaee/jakarta.ee/pull/1018/commits/38d21674c876af99947d29089eb395d2e4679853.  Does that look good?

**Review by ivargrimstad — APPROVED** [review]

File: content/committees/specification/tckprocess/index.md
Comment (edbratt) [reply content/committees/specification/tckprocess/index.md:435]: Once a TCK is released, it should be possible for an implementation to use it to qualify as compatible forever. In our experience -- adding excludes to the latest bug-fix (dot dot release) isn't usually a problem. The worst that might happen is other excludes might be included. (Jakarta EE doesn't currently have a notion of alternate tests.)

If support for a new Java SE version is added -- so long as it is fully a compatible upgrade-- again there should not be a problem posed by adding new excludes to the latest version. If there were some incompatibility introduced by a new SE version, then the update would not be a compatible update and would require a new version update.

Historically, exclude lists had been made available separately. If we run into problems that somehow prevent a release from being produced, we could  consider creating a dash release that only changes the exclude list (i.e. x.y.z-something). Date-strings could be useful for the 'something' text. If we resorted to this, we would need to start tracking the details of the Exclude list in the CCRs.

Bottom line -- once a TCK is released -- it should be available to vendors to verify compatibility of the original feature set -- forever. Implementers should anticipate that any changes would be delivered in the next higher dot dot release, even if that forces them to skip forward across dot dot releases.

**Comment by kwsutter** [comment]
> @kwsutter I made one more change to cover the possible TCK changes that we might later make for the removal of the Java Security Manager via [38d2167](https://github.com/jakartaee/jakarta.ee/commit/38d21674c876af99947d29089eb395d2e4679853). Does that look good?

Yes, this is even better.  Thanks!

**Review by kwsutter — APPROVED** [review]
LGTM!  Thanks, @scottmarlow!

**Comment by scottmarlow** [comment]
@chrisguindon could this pull request be merged please, as it is now marked approved.

**Comment by chrisguindon** [comment]
@scottmarlow  Thanks for the ping. The review was missing a +1 from @edbratt!

Never hesitate to ping us if something is ready for us to merge as you did. It's not always obvious to us that these types of PR are ready to go.

@ivargrimstad is a committer on the Jakarta.ee website, he can merge these content update if we are holding things up!

**Pull request merged**
